src/hotspot/share/prims/jvmtiTagMapTable.cpp

```diff
@@ -37,190 +37,103 @@
-  for (int i = 0; i < table_size(); ++i) {
-    for (JvmtiTagMapEntry* m = bucket(i); m != NULL;) {
-      JvmtiTagMapEntry* entry = m;
-      // read next before freeing.
-      m = m->next();
-      free_entry(entry);
-    }
-    JvmtiTagMapEntry** p = bucket_addr(i);
-    *p = NULL; // clear out buckets.
+JvmtiTagMapEntry::JvmtiTagMapEntry(oop obj): _obj(obj){}
+
+JvmtiTagMapEntry::JvmtiTagMapEntry(const JvmtiTagMapEntry& src) {
+   // move object into WeakHandle when copying into the table
+   assert(src._obj != nullptr, "must be set");
+   _wh = WeakHandle(JvmtiExport::weak_tag_storage(), src._obj);
+   _obj = nullptr;
+ }
+
+JvmtiTagMapEntry::~JvmtiTagMapEntry(){
+   // If obj is set null it out, this is called for stack object on lookup,
+   // and it should not have a WeakHandle created for it yet.
+   if (_obj != nullptr) {
+     _obj = nullptr;
+     assert(_wh.is_null(), "WeakHandle should be null");
+   } else {
+     _wh.release(JvmtiExport::weak_tag_storage());
   }
-  assert(number_of_entries() == 0, "should have removed all entries");
-}
-
-JvmtiTagMapTable::~JvmtiTagMapTable() {
-  clear();
-  // base class ~BasicHashtable deallocates the buckets.
 }
 
-// Entries are C_Heap allocated
-JvmtiTagMapEntry* JvmtiTagMapTable::new_entry(unsigned int hash, WeakHandle w, jlong tag) {
-  JvmtiTagMapEntry* entry = (JvmtiTagMapEntry*)Hashtable<WeakHandle, mtServiceability>::new_entry(hash, w);
-  entry->set_tag(tag);
-  return entry;
+oop JvmtiTagMapEntry::object() const {
+  assert(_obj == nullptr, "Must have a handle and not object");
+  return _wh.resolve();
 }
 
-void JvmtiTagMapTable::free_entry(JvmtiTagMapEntry* entry) {
-  entry->literal().release(JvmtiExport::weak_tag_storage()); // release to OopStorage
-  BasicHashtable<mtServiceability>::free_entry(entry);
+oop JvmtiTagMapEntry::object_no_keepalive() const {
+  assert(_obj == nullptr, "Must have a handle and not object");
+  return _wh.peek();
 }
 
-unsigned int JvmtiTagMapTable::compute_hash(oop obj) {
-  assert(obj != NULL, "obj is null");
-  return obj->identity_hash();
+JvmtiTagMapTable::JvmtiTagMapTable()
+  :_rrht_table(Constants::_table_size){
 }
 
-JvmtiTagMapEntry* JvmtiTagMapTable::find(int index, unsigned int hash, oop obj) {
-  assert(obj != NULL, "Cannot search for a NULL object");
-
-  for (JvmtiTagMapEntry* p = bucket(index); p != NULL; p = p->next()) {
-    if (p->hash() == hash) {
-
-      // Peek the object to check if it is the right target.
-      oop target = p->object_no_keepalive();
-
-      // The obj is in the table as a target already
-      if (target == obj) {
-        ResourceMark rm;
-        log_trace(jvmti, table)("JvmtiTagMap entry found for %s index %d",
-                                obj->print_value_string(), index);
-        // The object() accessor makes sure the target object is kept alive before
-        // leaking out.
-        (void)p->object();
-        return p;
-      }
+void JvmtiTagMapTable::clear() {
+  struct RemoveAll{
```

Comment:
Nit: space before {

Stylistically I'm not sure we define local structs like this when defining a "closure" object. ??

Reply:
I have seen these local structs in some places in the code (e.g., in concurrentHashTable.hpp), therefore I used it this way. 
Any preference?